### PR TITLE
Throw exception if init with wrong dimension

### LINKED ISSUE #7: getNearest fails silently when dimensionality of ANNIndex doesn't match dimensionality of loaded tree object
If I make a tree in python with dimensionality 10, like so:
```python
from annoy import AnnoyIndex
import random

f = 10  # specifying dimensionality 10
t = AnnoyIndex(f)
for i in range(1000):
    v = [random.gauss(0, 1) for z in range(f)]
    t.add_item(i, v)

t.build(5)
t.save('/Users/ZacP/test.ann')
```
And then try to load that tree in a scio job, run in the REPL using `sbt repl/run`:

```scala
import com.spotify.scio.values.DistCache
import org.joda.time.format.DateTimeFormat
import scala.collection.JavaConversions._
import com.spotify.annoy.{ANNIndex, AnnoyIndex}

val annTree: DistCache[AnnoyIndex] = sc.distCache {
      "/Users/ZacP/test.ann"
    } { f => new ANNIndex(11, f.getPath) }  // the "11" here specifies a dimensionality that is greater than what the tree actually has

val myCollection = 0 until 10

myCollection.flatMap {
  r =>
    val tree = annTree()
    tree.getNearest(List.fill(10)(0.0).map(_.toFloat).toArray, 10).toList
}
```
The result of the `flatMap()` stored in `output` is just an empty vector. If we make a new `ANNIndex` with dimensionality _less_ than what the underlying tree actually has, then it throws an `IndexOutOfBoundsException`. This is inconsistent behavior. I propose that there should be an error message telling the user that the dimensionality of the ANNIndex doesn't match that of the query vector, regardless of whether the query vector is shorter or longer than expected.

### PR DESCRIPTION
**Pull request by 5antelope**
Problem: 
Try to fix #7 

Solution:
When we call load() at in constructor, if (file_size % dimension) != 0, it indicates there is a mismatch between index file and passed in dimension

### CODE REVIEW

**Comment by erikbern** [comment]
is this going to work? i think you need to check that it's modulo the size of a vector, not just the number of dimensions

also note that this can have false negatives

**Comment by 5antelope** [comment]
Thanks @erikbern 

I added a unit test for it -- ANNIndexTest#testLoadFile: throw an exception when index dimension does not match with dimension in ANNIndex() constructor.

How about we check (file_size % NODE_SIZE) == 0?

I know that it will still introduce false negatives, but ideally the only way to fix that would be adding metadata in the header? The sanity checks here would help to reduce the chances of mistake :-)

**Comment by 5antelope** [comment]
Sorry, I misread the comment. 

Add a check in ANNIndex#getNearest() to make sure the vector passed in should always be same size as DIMENSION;

Also, I think the check in load(), although introduces some false positive, would be nice to have as well?

**Comment by 5antelope** [comment]
Bump -- can I have a quick review on the change? Thanks! :-)

**Comment by 5antelope** [comment]
Hi @erikbern, does the change look reasonable to you? Thanks!

**Comment by erikbern** [comment]
looks like you need to rebase!

**Comment by erikbern** [comment]
also i defer to @a1k0n to determine correctness

File: src/main/java/com/spotify/annoy/ANNIndex.java [hunk 1]
```diff
@@ -70,6 +69,11 @@ public ANNIndex(final int dimension,
   private void load(final String filename) throws IOException {
     memoryMappedFile = new RandomAccessFile(filename, "r");
     int fileSize = (int) memoryMappedFile.length();
+
+    if (fileSize % NODE_SIZE != 0) {
```
Comment (a1k0n) [inline src/main/java/com/spotify/annoy/ANNIndex.java]: This won't work anymore after the recent >2GB support merge.

**Review by a1k0n — COMMENTED** [review]
In general it looks reasonable to me and probably should have already had these exceptions, but yeah, you'll need to resolve conflicts.

File: src/main/java/com/spotify/annoy/ANNIndex.java (hunk 1 above)
Comment (5antelope) [reply src/main/java/com/spotify/annoy/ANNIndex.java]: Thanks @a1k0n! Could you link me to the commit for the change?

**Comment by 5antelope** [comment]
Hi @erikbern @a1k0n -- I resolved the conflicts and adjust logic based on 2GB change:

it looks for two cases:

- things fit into multiple buffers - then file_size % node_size should be 0
- if things cannot be a multiple of block_size, then both rest % node_size of (file_size - rest) % node_size should be 0

Let me know if the change makes sense to you.

Thanks!

**Comment by 5antelope** [comment]
@a1k0n mind taking a look while you are free? Thanks! 😄

File: src/main/java/com/spotify/annoy/ANNIndex.java [hunk 2]
```diff
@@ -88,31 +88,37 @@ private void load(final String filename) throws IOException {
...
-        int k = annBuf.getInt(i);  // node[i].n_descendants
-        if (m == -1 || k == m) {
-          roots.add(index);
-          m = k;
-        } else {
-          process = false;
+    // Two valid relations between dimension and file size:
+    // 1) if everything fits into buffer, file size should be a multiple of NODE_SIZE (rest == 0);
+    // 2) otherwise, both rest of (file_size - rest) should be a multiple of NODE_SIZE.
+    if (rest == 0 || rest % NODE_SIZE == 0 && (fileSize - rest) % NODE_SIZE == 0) {
```
Comment (a1k0n) [inline src/main/java/com/spotify/annoy/ANNIndex.java]: I would invert this if clause and throw the exception first, and then not have to indent everything else.

Comment (5antelope) [reply src/main/java/com/spotify/annoy/ANNIndex.java]: sounds good. done!

**Review by a1k0n — APPROVED** [review]
👍

**Comment by 5antelope** [comment]
@erikbern ready to land 😎

**Pull request merged**